pkg/virtctl/create/preference/preference.go

```diff
@@ -0,0 +1,202 @@
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package preference
+
+import (
+	"github.com/spf13/cobra"
+	"k8s.io/apimachinery/pkg/util/rand"
+	"sigs.k8s.io/yaml"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	instancetypev1alpha2 "kubevirt.io/api/instancetype/v1alpha2"
+
+	"kubevirt.io/kubevirt/pkg/virtctl/create/params"
+)
+
+const (
+	Preference = "preference"
+
+	CPUFlag        = "cpu"
+	VolumeFlag     = "volume"
+	MachineFlag    = "machine"
+	NameFlag       = "name"
+	NamespacedFlag = "namespaced"
+
+	CPUValueErr = "CPU musth have a value of preferCores, preferSockets or preferThreads"
+)
+
+type createPreference struct {
+	name       string
+	namespaced bool
+	cpu        string
+	machine    string
+	volume     string
+}
+
+type optionFn func(*createPreference, *instancetypev1alpha2.VirtualMachinePreferenceSpec) error
+
+var optFns = map[string]optionFn{
+	VolumeFlag:  withVolumes,
+	MachineFlag: withMachine,
+	CPUFlag:     withCPU,
+}
+
+func NewCommand() *cobra.Command {
+	c := createPreference{}
+	cmd := &cobra.Command{
+		Use:     Preference,
+		Short:   "Create a VirtualMachinePreference or VirtualMachineClusterPreference manifest.",
+		Example: c.usage(),
+		RunE: func(cmd *cobra.Command, _ []string) error {
+			return c.run(cmd)
+		},
+	}
+	cmd.Flags().BoolVar(&c.namespaced, NamespacedFlag, c.namespaced, "Specify if VirtualMachinePreference should be created. By default VirtualMachineClusterPreference is created.")
+	cmd.Flags().StringVar(&c.name, NameFlag, c.name, "Specify the name of the Preference.")
+	cmd.Flags().StringVar(&c.volume, VolumeFlag, c.volume, "Defines the preferred storageClass")
+	cmd.Flags().StringVar(&c.machine, MachineFlag, c.machine, "Defines the preferred machine type to use.")
+	cmd.Flags().StringVar(&c.cpu, CPUFlag, c.cpu, "Defines the preferred guest visible CPU topology.")
+
+	return cmd
+}
+
+func (c *createPreference) setDefaults(cmd *cobra.Command) {
+	if cmd.Flags().Changed(NameFlag) {
+		return
+	}
+
+	if c.namespaced {
+		c.name = "preference-" + rand.String(5)
+	} else {
+		c.name = "clusterpreference-" + rand.String(5)
+	}
+}
+
+func withVolumes(c *createPreference, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec) error {
+	preferenceSpec.Volumes = &instancetypev1alpha2.VolumePreferences{
+		PreferredStorageClassName: c.volume,
+	}
+	return nil
+}
+
+func withMachine(c *createPreference, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec) error {
+	preferenceSpec.Machine = &instancetypev1alpha2.MachinePreferences{
+		PreferredMachineType: c.machine,
+	}
+	return nil
+}
+
+func withCPU(c *createPreference, preferenceSpec *instancetypev1alpha2.VirtualMachinePreferenceSpec) error {
+	if c.cpu != string(instancetypev1alpha2.PreferCores) &&
+		c.cpu != string(instancetypev1alpha2.PreferSockets) &&
+		c.cpu != string(instancetypev1alpha2.PreferThreads) {
+		return params.FlagErr(CPUFlag, CPUValueErr)
+	}
```

Comment:
I also wonder if we could get away with stashing logic like this alongside the constants? Something for a follow up however.

https://github.com/kubevirt/kubevirt/blob/91a4ac2114873d3102a558372920327c0c0ccd01/staging/src/kubevirt.io/api/instancetype/v1alpha2/types.go#L253-L263